src/main/java/uk/gov/ch/developer/docs/controller/developer/UserCallbackController.java

```diff
@@ -41,22 +52,90 @@ public String getCallback(@RequestParam("state") String state,
+
+        if (userProfileResponse != null) {
+
+            LOGGER.debug("Processing Response");
+            setUserProfile(signInData, userProfileResponse);
+            signInData.put(SessionKeys.SIGNED_IN.getKey(), Integer.valueOf(1));
+            chSession.getData().put(SessionKeys.SIGN_IN_INFO.getKey(), signInData);
+
+        }
+
+        LOGGER.debug("Redirecting...");
 
         return ("redirect:" + identityProvider.getRedirectUriPage());// TODO redirect back to page
                                                                      // where sign-in was initiated
     }
 
+    private UserProfileResponse getUserProfile(Map<String, Object> signInData, WebClient webClient,
+            Mono<OAuthToken> oauthTokenResponse) {
+        LOGGER.debug("Getting User Profile....");
+
+        OAuthToken oauthToken = oauthTokenResponse.block();
+
+        return requestUserProfile(signInData, webClient, oauthToken);
+    }
+
+    private UserProfileResponse requestUserProfile(Map<String, Object> signInData, WebClient webClient,
+            OAuthToken tokenResponse) {
+        setAccessToken(signInData, tokenResponse);
+        return webClient.get().uri(URI.create(identityProvider.getProfileUrl()))
+                .headers(h -> h.setBearerAuth(tokenResponse.getToken()))
+                .accept(MediaType.APPLICATION_JSON).exchange()
+                .flatMap(response -> response.bodyToMono(UserProfileResponse.class))
+                .block();
+    }
+
+    private Mono<OAuthToken> getOAuthToken(String code, WebClient webClient) {
+        LOGGER.debug("Getting OAuth Token....");
+        return webClient.post().uri(URI.create(identityProvider.getTokenUrl()))
+                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+                .accept(MediaType.APPLICATION_JSON).body(BodyInserters
+                        .fromPublisher(Mono.just(getPostRequestBody(code)), String.class))
+                .retrieve()
+                .bodyToMono(OAuthToken.class);
+    }
+
+    private void setUserProfile(Map<String, Object> signInData, UserProfileResponse userProfileResponse) {
+        LOGGER.debug("Setting User Profile...");
+        Map<String, Object> userProfileData = new HashMap<>();
+
+        userProfileData.put(SessionKeys.EMAIL.getKey(), userProfileResponse.getEmail());
+        userProfileData.put(SessionKeys.USER_ID.getKey(), userProfileResponse.getId());
+        userProfileData.put(SessionKeys.LOCALE.getKey(), userProfileResponse.getLocale());
+        userProfileData.put(SessionKeys.SCOPE.getKey(), userProfileResponse.getScope());
+        userProfileData.put(SessionKeys.FORENAME.getKey(), userProfileResponse.getForename());
+        userProfileData.put(SessionKeys.SURNAME.getKey(), userProfileResponse.getSurname());
+        userProfileData.put(SessionKeys.PERMISSIONS.getKey(), userProfileResponse.getPermissions());
+
```

Comment:
Why not unwrap the UserProfile response to a single UserProfile entity and put that in, with a single key. Then new properties aren't missed, and `UserProfile` won't be fragmented. Consumers will be able to take the UserProfile directly, without rebuilding it field by field.

Reply:
As discussed earlier on - this was done in order to comply with the java-session-handler and how it unwraps and builds the user profile. Therefore it was necessary to add each field to the map here, which is then read field by field and set by the java-session-handler.